tests/blackbox.py

```diff
@@ -131,21 +131,25 @@ def decorator_blackbox(func: SubroutineFnWrapper | ABIReturnSubroutine):
 
 # ---- API ---- #
 
+O = TypeVar("O")
 
-def mode_to_execution_mode(mode: Mode) -> blackbox.ExecutionMode:
-    if mode == Mode.Application:
-        return blackbox.ExecutionMode.Application
-    if mode == Mode.Signature:
-        return blackbox.ExecutionMode.Signature
 
-    raise Exception(f"Unknown mode {mode} of type {type(mode)}")
+def _match_mode(mode: Mode, app: Callable[[], O], signature: Callable[[], O]) -> O:
```

Comment:
This is clever, but perhaps too clever? It took me quite a while to grok what was happening and I'm not sure such simple logic really requires all this fancy footwork. I think a little violation of D.R.Y. is ok in this case.

If you feel strongly about it then consider renaming `signature` and `app` resp. to `app_case_lambda` and `signature_case_lambda`. 

I think I see now why you were thinking about partial. I believe you can achieve similar functionality with a little class such as the following (but haven't tested it and not convinced it has any advantages):

```python
@dataclass
class _mode_matcher:
    app_case_lambda: Callable[[], Output]
    signature_case_lambda: Callable[[], Output]

   def __call__(self, mode: Mode) -> Output:
       ... match similar to the as above ...
```
Usage:
```python
>>> mode_to_execution_mode = _match_mode(lambda: blackbox.ExecutionMode.Application, lambda: blackbox.ExecutionMode.Signature)
...
>>> graviton_mode = mode_to_execution_mode(mode)
# inside `PyTealDryRunExecutor`:
    dryrun_sequence_runner = _match_mode(lambda: compileTeal...)
# etc...
```

Reply:
@tzaffi A few thoughts here:

> This is clever, but perhaps too clever?

I went down the shown path because of the need to repeat the pattern match N-times over.  Due to lack of exhaustiveness checking, my primary concern is ensuring the unsupported case exists.  The structure adds a little overhead in exchange for guaranteeing correct usage.

I can see the approach is probably less Pythonic and more Scala-like.  It's not my intent to make comprehension opaque.

>  achieve similar functionality with a little class 

https://github.com/algorand/pyteal/pull/338/commits/685ae794536cb6d83ccb8fab1ff3d21b69bbdb89 implements the suggestion.

My 2c (biased by Scala background) is that the class approach adds more indirection than the function.  That said, my primary desire to is stay within Pythonic norms:
* _If_ you find the class approach easier to reason about and valuable enough to retain, I'm happy to keep as is.
* _If_ you're still on the edge, feel welcomed to tell me.  I'd rather remove the questionable functionality.